<a href="https://colab.research.google.com/github/praveen555/IMI-BIGDATA--CHALLENGE/blob/main/Name_Matching_KYC.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [3]:
import os
import pandas as pd
import re 
import matplotlib.pyplot as plt
from datetime import datetime, date
import numpy as np
from google.colab import drive
drive.mount('/content/drive')
path='/content/drive/MyDrive/Colab Notebooks/IMI_BigData_Challenge/Raw data files'
path_2='/content/drive/MyDrive/Colab Notebooks/IMI_BigData_Challenge/Task_1_Data'
path_edge=os.path.join(path,'UofT_edges.csv')
path_nodes=os.path.join(path,'UofT_nodes.csv')
path_occupation_risk=os.path.join(path,'UofT_occupation_risk.csv')
path_name_list='/content/drive/MyDrive/Colab Notebooks/IMI_BigData_Challenge/name_list.csv'
path_clean_customer=os.path.join(path_2,'clean_customer_names.csv')
path_clean_sanctions=os.path.join(path_2,'clean_sanction_names.csv')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [9]:
sanction_names=pd.read_csv(path_clean_sanctions)
customer_names=pd.read_csv(path_clean_customer)

In [10]:
sanction_names

Unnamed: 0                             NAME  \
0                0                   Kolyvanov Egor   
1                1          Shipov Sergei Yurievich   
2                2         Egorov Ivan Mikhailovich   
3                3   Goreslavsky Alexey Sergeyevich   
4                4   Samoilova Natalya Vladimirovna   
...            ...                              ...   
247138      247138          ISABEL PATRICIA CARRIZO   
247139      247139            MARCELO DANIEL ROMERO   
247140      247140             GESAWORLD GROUP S.A.   
247141      247141  GESAWORLD USA LIMITED LIABILITY   
247142      247142                     SEYDOU IDANI   

                               aliases    BIRTH_DT  \
0                       Колыванов Егор  1980-11-15   
1                 Шипов Сергей Юрьевич  1966-04-17   
2               Егоров Иван Михайлович  1961-01-21   
3       Гореславский Алексей Сергеевич  1977-07-13   
4       Самойлова Наталья Владимировна  1987-06-24   
...                                ...         ...   
247138                             NaN         NaN   
247139                             NaN         NaN   
247140                             NaN         NaN   
247141                             NaN         NaN   
247142                             NaN         NaN   

                          cleaned_names  
0                        kolyvanov egor  
1               shipov sergei yurievich  
2              egorov ivan mikhailovich  
3        goreslavsky alexey sergeyevich  
4        samoilova natalya vladimirovna  
...                                 ...  
247138          isabel patricia carrizo  
247139            marcelo daniel romero  
247140             gesaworld group s a   
247141  gesaworld usa limited liability  
247142                     seydou idani  

[247143 rows x 5 columns]

In [11]:
customer_names

Unnamed: 0                            NAME    BIRTH_DT  \
0                0    Young, Marie Mildren Coleman  1981-09-01   
1                1                Mark Stupar Lecy  1994-02-21   
2                2      Dean Glasper Wendel Reeves  1962-11-16   
3                3             Hulsey, Linda Rauth  1998-06-20   
4                4      Carolyn Washington Roberts  1942-01-24   
...            ...                             ...         ...   
999335      999995               Bret* Dove Gainey  1939-01-16   
999336      999996       Anthony Ray Montano Gomez  1970-12-13   
999337      999997  Bernadette Schofield Pace Mile  1958-09-09   
999338      999998         Wasson, Russell Terrell  1983-09-15   
999339      999999                Daniel Mckeighan  1962-01-23   

                         cleaned_names  
0         young  marie mildren coleman  
1                     mark stupar lecy  
2           dean glasper wendel reeves  
3                  hulsey  linda rauth  
4           carolyn washington roberts  
...                                ...  
999335               bret  dove gainey  
999336       anthony ray montano gomez  
999337  bernadette schofield pace mile  
999338         wasson  russell terrell  
999339                daniel mckeighan  

[999340 rows x 4 columns]

In [15]:
df=pd.merge(customer_names,sanction_names,on='BIRTH_DT',validate='one_to_many')

MergeError: ignored

In [13]:
df

Unnamed: 0_x                        NAME_x    BIRTH_DT  \
0                   0  Young, Marie Mildren Coleman  1981-09-01   
1                   0  Young, Marie Mildren Coleman  1981-09-01   
2                   0  Young, Marie Mildren Coleman  1981-09-01   
3                   0  Young, Marie Mildren Coleman  1981-09-01   
4                   0  Young, Marie Mildren Coleman  1981-09-01   
...               ...                           ...         ...   
5001046        994339               June Vick Fultz  1921-04-21   
5001047        994339               June Vick Fultz  1921-04-21   
5001048        998751            Nick Tamayo Turner  1925-08-31   
5001049        998751            Nick Tamayo Turner  1925-08-31   
5001050        998751            Nick Tamayo Turner  1925-08-31   

                      cleaned_names_x  Unnamed: 0_y               NAME_y  \
0        young  marie mildren coleman         31879        Adrian CUSIAC   
1        young  marie mildren coleman         56731    PETROVIC, Zelimir   
2        young  marie mildren coleman        115690  José Ignacio Paliza   
3        young  marie mildren coleman        116000    Inger Lise Hansen   
4        young  marie mildren coleman        123005          Matt Warman   
...                               ...           ...                  ...   
5001046               june vick fultz          8278      Pierre Teuscher   
5001047               june vick fultz        121063      Nicolaus Dreyer   
5001048            nick tamayo turner         65492       Paolo Tringali   
5001049            nick tamayo turner        160687        Dewey L. Hill   
5001050            nick tamayo turner        188003      Shirley Walters   

                                                   aliases  \
0                                                      NaN   
1                                         Zelimir Petrovic   
2        Jose Ignacio Paliza;Jose Paliza;José Ignacio P...   
3                                                      NaN   
4        Matthew Robert Warman;Warman, Matt;مات وارمان;...   
...                                                    ...   
5001046                                   Teuscher, Pierre   
5001047              Dreyer!;Nikolaus Dreyer;نيكولاوس درير   
5001048                                                NaN   
5001049  Dewey Lewis Hill;ديوى ال. هيل;ديوي إل. هيل;دیو...   
5001050    Mary Shirley Walters;Mary Walters;Walters, Mary   

             cleaned_names_y  
0              adrian cusiac  
1          petrovic  zelimir  
2        jos  ignacio paliza  
3          inger lise hansen  
4                matt warman  
...                      ...  
5001046      pierre teuscher  
5001047      nicolaus dreyer  
5001048       paolo tringali  
5001049        dewey l  hill  
5001050      shirley walters  

[5001051 rows x 8 columns]

In [16]:
pip install fuzzywuzzy


Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [17]:
pip install python-levenshtein

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 174 kB 5.1 MB/s 
     |████████████████████████████████| 2.2 MB 19.6 MB/s 


In [18]:
from fuzzywuzzy import fuzz
from fuzzywuzzy import process

In [20]:
type(x)

int

In [25]:
def fuzzyscore(col):
  return fuzz.token_set_ratio(col['cleaned_names_x'],col['cleaned_names_y'])



In [27]:
df['score']=df.apply(fuzzyscore,axis=1)

In [53]:
high_risk_names=df[df['score']>75]

In [54]:
high_risk_names.shape

(119, 9)

In [56]:
high_risk_names.sort_values('score',ascending=False,inplace=True)

/usr/local/lib/python3.8/dist-packages/pandas/util/_decorators.py:311: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return func(*args, **kwargs)


In [57]:
high_risk_names

Unnamed: 0_x                        NAME_x    BIRTH_DT  \
3717284        983455           Samantha Lewthwaite  1983-12-05   
4436039        142376  David Stender Morris Helmick  1955-12-30   
2854462        983459          Paul Franklin Watson  1950-12-02   
2597588        723832                             s  1959-12-02   
1028987        125672       Alison Merrilla Redford  1965-03-07   
...               ...                           ...         ...   
2830844        468430              Samantha Patrick  1964-06-18   
3376091        422149           Maria CASE Thompson  1957-08-12   
3376021        197034         Kenneth Lane Thompson  1957-08-12   
1327525        108284     Nix, Nicholas Allen White  1955-04-22   
1627013          8997    Debordah Christopher Lopez  1950-09-21   

                      cleaned_names_x  Unnamed: 0_y  \
3717284           samantha lewthwaite        101394   
4436039  david stender morris helmick        160269   
2854462          paul franklin watson        122124   
2597588                             s        166133   
1028987       alison merrilla redford        152058   
...                               ...           ...   
2830844              samantha patrick        118072   
3376091           maria case thompson        192560   
3376021         kenneth lane thompson        192560   
1327525     nix  nicholas allen white        184615   
1627013    debordah christopher lopez        159023   

                             NAME_y  \
3717284         Samantha Lewthwaite   
4436039                David Morris   
2854462        PAUL FRANKLIN WATSON   
2597588  Anuario Luis Herrera Solís   
1028987              Alison Redford   
...                             ...   
2830844               Samantha Dunn   
3376091               Mike Thompson   
3376021               Mike Thompson   
1327525               Nicholas Yang   
1627013          Christopher Graham   

                                                   aliases  \
3717284  Bílá vdova;SAMANTHA LOUISE LEWTHWAITE;Samantha...   
4436039        David Charles Morris;ديفيد موريس;ডেভিড মরিস   
2854462  Paul Franklin Watson;Paul Watson;Пол Ватсон;По...   
2597588  Anuario Luis Herrera Solis;انواريو لويس هيريرا...   
1028987  Alison Merrilla Redford;Άλισον Ρέντφορντ;Елісо...   
...                                                    ...   
2830844                                                NaN   
3376091                 Michael Lowry Thompson;মাইক টম্পসন   
3376021                 Michael Lowry Thompson;মাইক টম্পসন   
1327525  Nicholas Yang Wei-hsiung;Yang Wei-hsiung;Yang ...   
1627013  Christopher Sidney Matthew Graham;Кристофер Гр...   

                    cleaned_names_y  score  
3717284         samantha lewthwaite    100  
4436039                david morris    100  
2854462        paul franklin watson    100  
2597588  anuario luis herrera sol s    100  
1028987              alison redford    100  
...                             ...    ...  
2830844               samantha dunn     76  
3376091               mike thompson     76  
3376021               mike thompson     76  
1327525               nicholas yang     76  
1627013          christopher graham     76  

[119 rows x 9 columns]

## Since only 119 names are present double check manually and see. I guess they told they want total 50 names but more is also fine ?? 


In [58]:
high_risk_names.to_csv('high_risk_customer_names.csv')

## Further Approaches 

1. Check with aliases to improve and filter.
2. Text conversion ? 
3. Other ideas any ?? 